_drafts/heartbeats.md

```diff
@@ -0,0 +1,963 @@
+still a challenge. Another unfortunate truth about binary logging is that a
+developer can no longer just connect to a UART and understand the logs.
+**Scripts are required**.
+
+### Heartbeats
+
+A device heartbeat can be thought of as a periodic "pulse" sent from a device to
+a monitoring service. The presence of the heartbeat alone can tell you whether a
+device is functional or online, but it can also store device vitals and
+statistics aggregated from the last time period. Devices that have a constant
+connection to the Internet and either a powered connection or a large battery
+might send a heartbeat every minute, and power-optimized devices might send one
+every hour or each day.
+
+Heartbeats differ from all of the other types mentioned above because the data
+is pre-aggregated on the device. Instead of sending raw log data and having to
+do stream processing on the logs to pull out metric data, we store the metric
+directly on the device!
+
+> The below examples use the heartbeat library I mentioned
+> [here](#heartbeat-library-example)
+
+Below, we increment the count of flash write errors seen on a device:
+
+```c
+void flash_write(...) {
+    ...
+    if (retval < 0) {
+        // Record there was an error
+        device_metrics_incr(kDeviceMetric_FlashError);
+    }
+    return retval;
+}
+```
+
+Here, we record that the device experienced a % drop in battery life over the
+last hour.
+
+<a name="battery-life-drain-metric"></a>
+
+```c
+void device_metrics_flush(void) {
+    ...
+    static int32_t s_prev_battery_pct;
+    const int32_t current_battery_pct = battery_get_pct();
+    const int32_t battery_delta = current_battery_pct - s_prev_battery_pct;
+    device_metrics_set(kDeviceMetricId_BatteryLifeDrain, battery_delta);
+}
+```
+
+<a name="wifi-ble-connectivity-metric"></a>
+
+Recall the [difficult-to-calculate metric](#wifi-ble-connectivity-metric) of
+Wi-Fi connected time per hour. It's easy when you count the time connected on
+the device!
+
+```c
+static uint32_t s_wifi_ticks;
+
+void wifi_connected_callback(void) {
+    // Start timer when Wi-Fi connects
+    device_metrics_timer_start(&s_wifi_ticks);
+}
+
+void wifi_disconnected_callback(void) {
+    // Stop timer when Wi-Fi connects
+    device_metrics_timer_end(kDeviceMetricWifiConnectedDuration
+                            &s_wifi_ticks);
+}
+```
+
+There is a critical rule to follow with heartbeats that make them work as well
+as they do. After the individual metrics are collected from the device and a
+heartbeat is packaged up, the **metric values aggregated from this interval are
+cleared**. This means the next heartbeat will contain entirely new data not
+based upon the previous interval.
+
+> More information on resetting metrics for each heartbeat can be found
+> [here](#resetting-data-for-each-heartbeat).
+
+**Pros:** Least amount of data sent, easily parsed by server, pre-aggregated for
```

Comment:
maybe mention somewhere what format the data is being published in to the server? I imagine one would face similar challenges to what you mention in binary logging around serialization strategies